pilot/pkg/networking/plugin/envoyfilter/envoyfilter.go

```diff
@@ -0,0 +1,301 @@
+				} else {
+					break
+				}
+			}
+		}
+	}
+}
+
+func insertNetworkFilter(filterChain *plugin.FilterChain, envoyFilter *networking.EnvoyFilter_Filter) {
+	filter := &listener.Filter{
+		Name:   envoyFilter.FilterName,
+		Config: envoyFilter.FilterConfig,
+	}
+
+	position := networking.EnvoyFilter_InsertPosition_FIRST
+	if envoyFilter.InsertPosition != nil {
+		position = envoyFilter.InsertPosition.Index
+	}
+
+	switch position {
+	case networking.EnvoyFilter_InsertPosition_FIRST, networking.EnvoyFilter_InsertPosition_BEFORE:
+		filterChain.TCP = append([]listener.Filter{*filter}, filterChain.TCP...)
+		if position == networking.EnvoyFilter_InsertPosition_BEFORE {
+			// bubble the filter to the right position scanning from beginning
+			for i := 1; i < len(filterChain.TCP); i++ {
+				if filterChain.TCP[i].Name != envoyFilter.InsertPosition.RelativeTo {
+					filterChain.TCP[i-1], filterChain.TCP[i] = filterChain.TCP[i], filterChain.TCP[i-1]
+				}
+			}
+		}
+	case networking.EnvoyFilter_InsertPosition_LAST, networking.EnvoyFilter_InsertPosition_AFTER:
+		filterChain.TCP = append(filterChain.TCP, *filter)
+		if position == networking.EnvoyFilter_InsertPosition_AFTER {
+			// bubble the filter to the right position scanning from end
+			for i := len(filterChain.TCP) - 2; i >= 0; i-- {
+				if filterChain.TCP[i].Name != envoyFilter.InsertPosition.RelativeTo {
+					filterChain.TCP[i+1], filterChain.TCP[i] = filterChain.TCP[i], filterChain.TCP[i+1]
+				}
+			}
+		}
+	}
+
+}
+
+func insertUserSpecifiedFilters(in *plugin.InputParams, mutable *plugin.MutableObjects, direction string) error {
+	filterCRD := getFilterForWorkload(in)
+	if filterCRD == nil {
+		return nil
+	}
+
+	listenerIPAddress := getListenerIPAddress(&mutable.Listener.Address)
+	if listenerIPAddress == nil {
+		log.Warnf("Failed to parse IP Address from plugin listener")
+	}
+
+	for _, f := range filterCRD.Filters {
+		if !listenerMatch(in, direction, listenerIPAddress, f.ListenerMatch) {
+			continue
+		}
+		if in.ListenerProtocol == plugin.ListenerProtocolHTTP && f.FilterType == networking.EnvoyFilter_Filter_HTTP {
+			// Insert into http connection manager
+			for cnum := range mutable.FilterChains {
+				insertHTTPFilter(&mutable.FilterChains[cnum], f)
+			}
+		} else {
+			// tcp listener with some opaque filter or http listener with some opaque filter
+			// We treat both as insert network filter X into network filter chain.
+			// We cannot insert a HTTP in filter in network filter chain. Even HTTP connection manager is a network filter
+			if f.FilterType == networking.EnvoyFilter_Filter_HTTP {
+				log.Warnf("Ignoring filter %s. Cannot insert HTTP filter in network filter chain", f.FilterName)
+				continue
+			}
+			for cnum := range mutable.FilterChains {
+				insertNetworkFilter(&mutable.FilterChains[cnum], f)
+			}
+		}
+	}
+	return nil
+}
+
+// OnOutboundListener implements the Callbacks interface method.
```

Comment:
since there's no need for forward declarations it's more common in go to put the calling functions before the called functions so top to bottom in the file is same as top to bottom in the call tree.